extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisor.java

```diff
@@ -212,14 +225,47 @@ protected void scheduleReporting(ScheduledExecutorService reportingExec)
     // not yet implemented, see issue #6739
   }
 
+
+  /**
+   * We try to parse the shard number of the shard ID, using a BigInteger because the Kinesis shard ID can be
+   * up to 128 characters. The shard number is used preferentially because it provides a fixed and easily predictable
+   * mapping from shard to task group number.
+   *
+   * If we can't parse the shard number from the ID, then we fall back to hashing the shard ID string.
```

Comment:
Why would this happen? If that might happen, wouldn't it be better to reduce the potential code paths and simplify things? If you mix approaches it becomes harder to figure out what was used to make the calculation.

Reply:
I don't think it can happen today, given how Kinesis names shards. It might happen in the future if they change the format of a shard identifier for some reason. I don't think that's likely to happen, since it would probably break a ton of stuff so Amazon would hesitate to do it. So just doing the BigInteger path only (and throwing an exception if they can't be parsed for some reason) is a risk but a small one. On the other hand it's zero risk to treat them as opaque strings.

I read the code in the PR as trying to have it both ways, a 'nice' behavior (assign shards round-robin based on shard number) but also zero risk of Amazon doing an unexpected change and breaking us.

Fwiw I think I would go with using the zero risk SHA path only — it would mean shard assignment that isn't as human-predictable, but it shortens up the code while still avoiding forward-compatibility risk, which seems worth it to me.

In the end I am inclined to leave it up to the author @jon-wei.

Reply:
Hm, I really like being able to compute the shard -> task group mapping mentally and the round-robin behavior, but there is some uncertainty there. 

The shardId format expected there is how Kinesis names the shards, and I don't see any way to rename shards in the API (https://docs.aws.amazon.com/kinesis/latest/APIReference/Welcome.html). 

That said, the API doesn't really define the format of the shard names, beyond the allowed character set and length, so I think I'll go with only the hash path given the feedback and that uncertainty.